File: laravel-backend/app/Http/Controllers/Api/RatesController.php
<?php

namespace App\Http\Controllers\Api;

use App\Services\GoldPriceHistory;
use App\Services\MetalRates;
use App\Services\SharedRates;
use App\Support\Pos;
use Illuminate\Http\Request;

class RatesController extends ApiController
{
    public function metalRates(Request $request)
    {
        if (!MetalRates::isConfigured()) return $this->fail('Live metal API is not configured.', 503);
        try {
            $currency = MetalRates::normalizeMetalCurrency((string) $request->query('currency', 'USD'));
            $rates = MetalRates::getLiveRates($currency);
            $tolaNpr = SharedRates::displayToNpr($rates['gold']['perTola'] ?? 0, $currency);
            $gramNpr = SharedRates::displayToNpr($rates['gold']['perGram'] ?? 0, $currency)
                ?: Pos::round2($tolaNpr / Pos::TOLA_GRAMS);
            if ($tolaNpr > 0) {
                try {
                    SharedRates::appendHistory([
                        'goldRatePerTola' => $tolaNpr, 'goldRatePerGram' => $gramNpr,
                        'priceMode' => 'api', 'localDate' => SharedRates::localDateStr(),
                    ]);
                } catch (\Throwable $err) {
                    // History save is best-effort (matches the Express behaviour).
                }
            }
            return response()->json($rates);
        } catch (\Throwable $err) {
            // The provider's raw error is logged, not returned: it names the
            // upstream host and sometimes the request that failed.
            \Illuminate\Support\Facades\Log::warning('metal-rates: ' . $err->getMessage());
            return $this->fail('Could not fetch live metal rates right now.', 502);
        }
    }

    /**
     * The shared row holds only the MARKET (api-mode) feed now. Each shop's
     * own manual rate history lives in its settings and is never shared, so
     * `priceMode=manual` here answers with nothing rather than with every
     * shop's private selling rate.
     */
    public function sharedGoldRates(Request $request)
    {
        $date = substr((string) ($request->query('date') ?: gmdate('Y-m-d')), 0, 10);
        if (!preg_match('/^\d{4}-\d{2}-\d{2}$/', $date)) $date = gmdate('Y-m-d');
        if ($request->query('priceMode') !== 'api') return response()->json(['ticks' => [], 'history' => []]);
        return response()->json(SharedRates::getForClient($date, 'api'));
    }

    /**
     * Only the administrator (or the cron job, via cronCapture) may write to
     * the shared feed. Any shop used to be able to POST unlimited ticks with
     * arbitrary dates — and with `saved:true` they were kept forever — which
     * both polluted every other shop's chart and made a cheap DoS.
     */
    public function appendTicks(Request $request)
    {
        $user = $request->attributes->get('authUser');
        if (!$user || !$user->is_admin) return $this->fail('Not found.', 404);
        $body = $request->json()->all();
        $ticks = is_array($body['ticks'] ?? null) ? $body['ticks'] : [];
        if (count($ticks) > 200) return $this->fail('Too many ticks in one request (max 200).');
        foreach ($ticks as $t) {
            if (!is_array($t)) return $this->fail('Each tick must be an object.');
            $rate = Pos::numOrNull($t['goldRatePerTola'] ?? null);
            if ($rate === null || $rate <= 0 || !is_finite($rate)) return $this->fail('Each tick needs a positive goldRatePerTola.');
            $t['priceMode'] = 'api';
        }
        $result = SharedRates::appendTicks($ticks);
        return response()->json(['ok' => true, 'count' => $result['count']]);
    }

    public function cronCapture(Request $request)
    {
        $secret = trim((string) env('CRON_SECRET', ''));
        $auth = (string) $request->header('Authorization', '');
        $headerSecret = (string) $request->header('x-cron-secret', '');
        $authorized = $secret !== '' && (hash_equals("Bearer {$secret}", $auth) || hash_equals($secret, $headerSecret));
        if (!$authorized) return $this->fail('Cron secret required.', 401);
        // The shared market-price history (gold_price_ticks) is the one the
        // charts read; the legacy shared_gold_rates blob is kept in step.
        try {
            $tick = GoldPriceHistory::capture();
        } catch (\Throwable $err) {
            return $this->fail($err->getMessage() ?: 'Could not fetch the live gold price.', 502);
        }
        $result = SharedRates::captureIfChanged(['currency' => $request->query('currency')]);
        $result['tick'] = $tick;
        return response()->json($result);
    }
}

File: backend/rates-service/src/main/kotlin/com/subarnapasal/rates/RatesController.kt
package com.subarnapasal.rates

import com.subarnapasal.common.ApiException
import com.subarnapasal.common.Doc
import com.subarnapasal.common.Pos
import com.subarnapasal.common.asDoc
import jakarta.servlet.http.HttpServletRequest
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
class RatesController(private val metal: MetalRatesService, private val shared: SharedRatesService) {

    @GetMapping("/api/metal-rates")
    fun metalRates(@RequestParam(required = false) currency: String?): Any {
        if (!metal.isConfigured()) throw ApiException("Live metal API is not configured.", 503)
        val code = MetalRatesService.normalizeMetalCurrency(currency ?: "USD")
        val rates = try {
            metal.getLiveRates(code)
        } catch (e: ApiException) {
            throw e
        } catch (e: Exception) {
            throw ApiException(e.message ?: "Could not fetch live metal rates.", 502)
        }
        val gold = (rates["gold"] as? Map<*, *>)
        val tolaNpr = shared.displayToNpr(gold?.get("perTola") ?: 0, code)
        val gramNpr = shared.displayToNpr(gold?.get("perGram") ?: 0, code).takeIf { it != 0.0 }
            ?: Pos.round2(tolaNpr / Pos.TOLA_GRAMS)
        if (tolaNpr > 0) {
            try {
                shared.appendHistory(linkedMapOf<String, Any?>(
                    "goldRatePerTola" to tolaNpr, "goldRatePerGram" to gramNpr,
                    "priceMode" to "api", "localDate" to shared.localDateStr(),
                ))
            } catch (e: Exception) { /* best-effort, matches Express behaviour */ }
        }
        return rates
    }

    @GetMapping("/api/shared/gold-rates")
    fun sharedGoldRates(
        @RequestParam(required = false) date: String?,
        @RequestParam(required = false) priceMode: String?,
    ): Any = shared.getForClient(
        (date ?: shared.localDateStr()).take(10),
        if (priceMode == "api") "api" else "manual",
    )

    @PostMapping("/api/shared/gold-rates/ticks")
    fun appendTicks(@RequestBody body: Doc): Any {
        val ticks = (body["ticks"] as? List<Any?>) ?: emptyList()
        val result = shared.appendTicks(ticks)
        return mapOf("ok" to true, "count" to result["count"])
    }

    @GetMapping("/api/cron/capture-gold-rate")
    fun cronCapture(request: HttpServletRequest, @RequestParam(required = false) currency: String?): Any {
        val secret = (System.getenv("CRON_SECRET") ?: "").trim()
        val auth = request.getHeader("Authorization") ?: ""
        val headerSecret = request.getHeader("x-cron-secret") ?: ""
        val authorized = secret.isNotEmpty() && (auth == "Bearer $secret" || headerSecret == secret)
        if (!authorized) throw ApiException("Cron secret required.", 401)
        return shared.captureIfChanged(currency)
    }

    // ── Internal endpoints for store-service (settings screen) ───────────

    @GetMapping("/internal/shared-rates")
    fun internalRead(): Any = shared.read()

    @PostMapping("/internal/shared-rates/history")
    fun internalAppendHistory(@RequestBody body: Doc): Any = shared.appendHistory(body)

    @DeleteMapping("/internal/shared-rates")
    fun internalClear(@RequestParam(required = false) priceMode: String?): Any =
        shared.clear(if (priceMode == "api") "api" else "manual")

    @PostMapping("/internal/shared-rates/write")
    fun internalWrite(@RequestBody body: Doc): Any {
        shared.write(body.asDoc())
        return mapOf("ok" to true, "kind" to "shared_rates")
    }
}
